### PR TITLE
Enhance Pip Dependency Parser to Handle Complex Commands and Edge Cases

### PR DESCRIPTION
**Pull request by OlawumiSalaam**
- [x]  Have you followed the guidelines in our [Contribution Guide](https://github.com/ersilia-os/ersilia/blob/master/CONTRIBUTING.md)
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

**Description**
This pull request improves the pip dependency parser to handle more complex pip install commands and edge cases. The current parser is limited to handling simple pip install commands that follow a basic structure; `pip install package_name==version`. But sometimes there are more complex cases of `pip install` commands that the current parser does not support, such as:

1. Installing from a GitHub repository:
 `pip install git+https://github.com/bpkelley/descriptastorus.git@d552f934757378a61dd1799cdb589a864032cd1b
`
The command includes a Git URL: `git+https://github.com/...` and a specific commit/branch: `@d552f934...`.
The current parser would fail here because it does not know how to handle URLs or recognize that there’s no "explicit version number" like ==1.2.3.

2. Using additional parameters
`pip install torchvision==0.17.1 --index-url https://download.pytorch.org/whl/cpu`
This command has the package name:` torchvision`, a version: `==0.17.1`, and additional parameter: `--index-url https://...`
Again, the parser would fail because it’s only designed to handle the basic structure, and it doesn’t account for any extra flags like` --index-url`
This PR addresses those limitations and enhances the parser to support the Parser to handle complex commands and edge Cases:

1. Support for Git-based Dependencies: The parser now correctly identifies pip install commands that install packages from GitHub repositories (e.g.,` git+https://github.com/...`).
It extracts and parses repository URLs and specific commit/branch details, allowing the parser to handle dependencies installed from source control.
2. Support for Additional Parameters: The parser has been updated to recognize and preserve additional pip flags, such as` --index-url` for custom package indices and other options like` --no-cache-dir`, `--upgrade`, etc.
These flags are now correctly parsed and maintained when reconstructing the command.
3. Generalized Parsing: The parser is no longer constrained to expecting exactly four components in the pip install command. Instead, it dynamically identifies the components, allowing for more flexibility.
It can now differentiate between:

- Main package details (name and version).
- URLs or paths (e.g., Git repositories).
- Flags and additional parameters.

**Changes made**

1. Refactoring: The pip dependency parsing logic has been refactored to handle edge cases, including Git-based dependencies and extra parameters.
2. New Parsing Logic: The parser now splits and processes pip install commands more flexibly, identifying different components based on their type.
3. Unit Tests: New unit tests have been added to verify the handling of edge cases like Git-based installations and additional flags. These tests ensure that the parser works correctly with the newly supported formats.

**Testing**
Updated the test file named `test_dockerfile_install_parser.py` in the tests folder. I added two dockerfile from erilia model [eos4e40](https://github.com/ersilia-os/eos4e40/blob/main/Dockerfile) and [eos4e41](https://github.com/ersilia-os/eos4e41/blob/main/Dockerfile). The dockerfiles are stored in the fixtures directory for testing the implementation of the parsers logic. 
All  tests passed successfully. This indicates that the test setup, imports, and functionality are working as expected. The output files are 
[test_dockerfile_install_parser_output_eos4e40_Dockerfile.txt](https://github.com/user-attachments/files/18131566/test_dockerfile_install_parser_output_eos4e40_Dockerfile.txt)
[test_dockerfile_install_parser_output_eos4e41_Dockerfile.txt](https://github.com/user-attachments/files/18131567/test_dockerfile_install_parser_output_eos4e41_Dockerfile.txt)

Is this pull request related to any open issue? If yes, replace issueID below with the issue ID
Yes,
Related to https://github.com/ersilia-os/ersilia-pack/issues/21

### CODE REVIEW

File: src/ersilia_pack/parsers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# __init__.py to mark this directory as a Python package
+#from ersilia-pack.parsers.install_parser import InstallParser
+#from parsers.dockerfile_install_parser import DockerfileInstallParser
```
Comment (DhanshreeA) [inline src/ersilia_pack/parsers/__init__.py:3]: We can remove the commented code.

File: src/ersilia_pack/parsers/dockerfile_install_parser.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import re
+#from .install_parser import InstallParser
```
Comment (DhanshreeA) [inline src/ersilia_pack/parsers/dockerfile_install_parser.py:2]: Remove the commented code.

File: src/ersilia_pack/parsers/yaml_install_parser.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          package: torchvision
+          version: 0.17.1
+          flags:
+            --index-url: https://download.pytorch.org/whl/cpu
+          url: git+https://github.com/pytorch/vision.git
+
+        Output: 
+        - "pip install torchvision==0.17.1 --index-url https://download.pytorch.org/whl/cpu"
+        - "pip install git+https://github.com/pytorch/vision.git"
+        """
```
Comment (DhanshreeA) [inline src/ersilia_pack/parsers/yaml_install_parser.py:54]: @OlawumiSalaam Can you share an example YAML? I don't think we need additional keys in the YAML. We still want to keep only `commands`, and then have list entries for that. So something like, 

```yaml
commands:
- ["pip", "torchvision", "0.17.1", "--index-url", "<url>"]
```

Asking a user to specify a yaml this detailed would not make for a good experience. 

A

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

**Comment by DhanshreeA** [comment]
@OlawumiSalaam can you also add an install.yml with edge-cases specified addressed in this PR as a test fixture?

**Comment by OlawumiSalaam** [comment]
@DhanshreeA okay, i will improve the implementation for improved usability. Can you point me to an example model that has `install.yml` file that I can use as test fixture. Thank you.

**Comment by DhanshreeA** [comment]
> @DhanshreeA okay, i will improve the implementation for improved usability. Can you point me to an example model that has `install.yml` file that I can use as test fixture. Thank you.

No model has an install.yml like this unfortunately, because this feature doesn't exist in Ersilia Pack yet.

**Comment by DhanshreeA** [comment]
@OlawumiSalaam I have merged @Malikbadmus' #39 PR which addresses this as needed. This is fine, because we need this feature quite urgently. I can fix the conflicts in your PR and push it to your branch - and then we can finalize your tests and merge this one.
